allennlp/modules/elmo.py

```diff
@@ -515,3 +565,63 @@ def forward(self,  # pylint: disable=arguments-differ
                 'activations': output_tensors,
                 'mask': mask,
         }
+
+    def create_cached_cnn_embeddings(self, tokens: List[str]) -> None:
+        """
+        Given a list of tokens, this method precomputes word representations
+        by running just the character convolutions and highway layers of elmo,
+        essentially creating uncontextual word vectors. On subsequent forward passes,
+        if all of the words in the batch have cached, the word ids are looked up from
+        an embedding, rather than being computed on the fly via the CNN encoder.
+
+        This function sets 3 attributes:
+
+        _word_embedding : ``torch.Tensor``
+            The word embedding for each word in the tokens passed to this method.
+        _bos_embedding : ``torch.Tensor``
+            The embedding for the BOS token.
+        _eos_embedding : ``torch.Tensor``
+            The embedding for the EOS token.
+
+        Parameters
+        ----------
+        tokens : ``List[str]``, required.
+            A list of tokens to precompute character convolutions for.
+        """
+        tokens = [ELMoCharacterMapper.bos_token, ELMoCharacterMapper.eos_token] + tokens
+        timesteps = 32
+        batch_size = 32
+        chunked_tokens = lazy_groups_of(iter(tokens), timesteps)
+
+        all_embeddings = []
+        device = get_device_of(next(self.parameters()))
+        for batch in lazy_groups_of(chunked_tokens, batch_size):
+            # Shape (batch_size, 32, 32, 50)
+            batched_tensor = batch_to_ids(batch)
+            # NOTE: This device check is for when a user calls this method having
+            # already placed the model on a device. If this is called in the
+            # constructor, it will probably happen on the CPU. This isn't too bad,
+            # because it's only a few convolutions and will likely be very fast.
+            if device >= 0:
+                batched_tensor = batched_tensor.cuda(device)
+            output = self._token_embedder(batched_tensor)
+            token_embedding = output["token_embedding"]
+            mask = output["mask"]
+            token_embedding, _ = remove_sentence_boundaries(token_embedding, mask)
+            all_embeddings.append(token_embedding.view(-1, token_embedding.size(-1)))
```

Comment:
Preference for not using `-1` in `.view`, for readability and for more obvious errors.